### PR TITLE
discoverd: in-memory heartbeating

### LINKED ISSUE #2073: High disk writes for discoverd
We've noted our single-node Flynn instance producing quite significant disk IO, including writes for discoverd. This is for a server that hosts a single Rails application, during a periods of minimal or no application utilisation.

We're frequently getting many 100s of K/s disk reads as you can see in this 1 minute sample:
```
12:00:02  9713 be/4 root        0.00 B/s   38.82 K/s  0.00 %  0.15 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:02  8619 be/4 root        0.00 B/s   38.82 K/s  0.00 %  0.14 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:03  9713 be/4 root        0.00 B/s  127.10 K/s  0.00 %  0.52 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:03 20732 be/4 root        0.00 B/s  157.91 K/s  0.00 %  0.41 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:03  8626 be/4 root        0.00 B/s   34.66 K/s  0.00 %  0.08 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:06 20732 be/4 root        0.00 B/s   42.68 K/s  0.00 %  0.28 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:06  8619 be/4 root        0.00 B/s   38.80 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:08 20732 be/4 root        0.00 B/s  201.57 K/s  0.00 %  1.27 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:08  8619 be/4 root        0.00 B/s  120.17 K/s  0.00 %  0.90 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:11  8619 be/4 root        0.00 B/s   38.88 K/s  0.00 %  0.15 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:11  9713 be/4 root        0.00 B/s   42.77 K/s  0.00 %  0.12 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:13  9713 be/4 root        0.00 B/s  206.23 K/s  0.00 %  0.53 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:13  8619 be/4 root        0.00 B/s  116.74 K/s  0.00 %  0.34 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:17 20732 be/4 root        0.00 B/s   38.86 K/s  0.00 %  0.11 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:17  8619 be/4 root        0.00 B/s   38.86 K/s  0.00 %  0.11 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:18  8619 be/4 root        0.00 B/s   42.70 K/s  0.00 %  0.10 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:19  8619 be/4 root        0.00 B/s  205.99 K/s  0.00 %  0.56 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:19 20732 be/4 root        0.00 B/s  116.60 K/s  0.00 %  0.35 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:22 20732 be/4 root        0.00 B/s   38.85 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:22  9713 be/4 root        0.00 B/s   38.85 K/s  0.00 %  0.12 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:23 20732 be/4 root        0.00 B/s   42.71 K/s  0.00 %  0.32 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:24 20732 be/4 root        0.00 B/s  210.01 K/s  0.00 %  0.63 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:24  9713 be/4 root        0.00 B/s  116.67 K/s  0.00 %  0.39 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:27  9713 be/4 root        0.00 B/s   38.83 K/s  0.00 %  0.46 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:27  8619 be/4 root        0.00 B/s   38.83 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:28  8619 be/4 root        0.00 B/s  120.42 K/s  0.00 %  0.94 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:28  9713 be/4 root        0.00 B/s   81.57 K/s  0.00 %  0.29 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:29  9713 be/4 root        0.00 B/s  124.32 K/s  0.00 %  0.35 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:29  8619 be/4 root        0.00 B/s   38.85 K/s  0.00 %  0.10 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:32  8619 be/4 root        0.00 B/s   38.68 K/s  0.00 %  0.14 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:32 20732 be/4 root        0.00 B/s   38.68 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:33 20732 be/4 root        0.00 B/s  190.78 K/s  0.00 %  4.45 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:33  8619 be/4 root        0.00 B/s  167.89 K/s  0.00 %  2.05 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:37 20732 be/4 root        0.00 B/s   38.50 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:37  9713 be/4 root        0.00 B/s   38.50 K/s  0.00 %  0.11 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:38  9713 be/4 root        0.00 B/s  233.32 K/s  0.00 %  0.62 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:38 17864 be/4 root        0.00 B/s  128.32 K/s  0.00 %  0.42 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:38 20732 be/4 root        0.00 B/s   42.77 K/s  0.00 %  0.12 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:41  9713 be/4 root        0.00 B/s   38.81 K/s  0.00 %  0.15 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:41  8619 be/4 root        0.00 B/s   38.81 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:42  9713 be/4 root        0.00 B/s   42.68 K/s  0.00 %  0.11 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:43  8619 be/4 root        0.00 B/s  283.71 K/s  0.00 %  0.78 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:43 20732 be/4 root        0.00 B/s   77.73 K/s  0.00 %  0.22 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:46 17864 be/4 root        0.00 B/s   38.87 K/s  0.00 %  0.14 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:46 20732 be/4 root        0.00 B/s   38.87 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:48 20732 be/4 root        0.00 B/s   42.75 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:49  8619 be/4 root        0.00 B/s  237.17 K/s  0.00 %  3.81 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:49 20732 be/4 root        0.00 B/s  124.42 K/s  0.00 %  0.76 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:52  8619 be/4 root        0.00 B/s   38.81 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:52 17864 be/4 root        0.00 B/s   38.81 K/s  0.00 %  0.11 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:53  8619 be/4 root        0.00 B/s   42.69 K/s  0.00 %  0.13 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:54  8619 be/4 root        0.00 B/s  283.43 K/s  0.00 %  0.74 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:54 17864 be/4 root        0.00 B/s  116.48 K/s  0.00 %  0.35 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:57 20732 be/4 root        0.00 B/s   38.85 K/s  0.00 %  0.12 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:57 17864 be/4 root        0.00 B/s   38.85 K/s  0.00 %  0.12 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:58 17864 be/4 root        0.00 B/s   42.74 K/s  0.00 %  0.89 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:59 17864 be/4 root        0.00 B/s  283.76 K/s  0.00 %  1.22 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
12:00:59 20732 be/4 root        0.00 B/s  116.61 K/s  0.00 %  0.78 % discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
```

The reads appear to be spread across multuple discoverd threads, with 12 threads running at the time:
```
root      8619  8613  8619  0   12 Oct17 ?        00:10:49 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8622  0   12 Oct17 ?        00:21:11 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8623  0   12 Oct17 ?        00:00:00 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8624  0   12 Oct17 ?        00:13:56 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8625  0   12 Oct17 ?        00:09:59 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8626  0   12 Oct17 ?        00:10:40 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  8646  0   12 Oct17 ?        00:13:11 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613  9713  0   12 Oct17 ?        00:09:16 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613 13013  0   12 Oct17 ?        00:10:02 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613 17864  0   12 Oct17 ?        00:15:23 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613 25285  0   12 Oct18 ?        00:08:21 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
root      8619  8613 20732  0   12 Nov01 ?        00:01:44 /bin/discoverd -data-dir=/data -host=10.50.0.61 -peers=10.50.0.61:1110, -raft-addr=:1110 -http-addr=:1111 -notify=http://10.50.0.61:1113/host/discoverd -wait-net-dns=true
```

### LINKED ISSUE #2158: discoverd: Heartbeating
## Overview

Currently `discoverd` can experience high write load because instances heartbeat their status every few seconds and this heartbeat then goes through raft consensus. Because it goes through consensus, all `discoverd` instances are kept in sync however this has a high cost.

Jonathan had mentioned moving heartbeats to a separate in-memory raft instance to avoid disk I/O, however, I don't think we need the strong consistency that raft provides. Also, raft probably won't do well to recover if a quorum of nodes loses all their data.

## Proposed Change

Instead of moving to another raft cluster, I propose that `discoverd` simply forwards heartbeats to the other nodes in the cluster. The state can be kept in-memory and is only used when a node is leader. If two separate heartbeats come in for the same instance, it is easy to merge them -- simply use the higher timestamp. In the event of a crash, the node simply waits for new heartbeats to come in before it can change leadership for a service.

We can piggyback onto the TCP connection that already exists by multiplexing it. Consul does this and I've written a TCP multiplexer before and it works well. In fact, we could also multiplex the HTTP traffic so it all goes over the same port.

### LINKED ISSUE #2168: various issues git push hangs, flynn create causes unknown_error: Something went wrong...
I'm working with a 3 node cluster in Azure. Installing flynn was a breeze; its great to be a able to deploy my own PaaS quickly and easily.

Everything worked fine for the first few pushes, but I'm now getting various intermittent errors, such as:

* flynn create in new repos fails intermittently with `unknown_error: Something went wrong`.
* flynn log hangs or returns `GET https://controller.qj13.flynnhub.com/apps/kookaburas-amuse-caradoc/log?follow=true: httpclient: raw req: unexpected status 503`
* git push hangs on certain repos

repairing the cluster sometimes seems to improve things temporarily (but difficult to tell), here is the output of `flynn-host fix --min-hosts 3`:
```
INFO[11-19|16:48:30] found expected hosts                     n=3
INFO[11-19|16:48:30] ensuring discoverd is running on all hosts 
INFO[11-19|16:48:30] checking flannel 
INFO[11-19|16:48:30] flannel looks good 
INFO[11-19|16:48:30] waiting for discoverd to be available 
INFO[11-19|16:48:30] checking for running controller API 
INFO[11-19|16:48:30] found running controller API instances   n=2
INFO[11-19|16:48:30] found controller instance, checking critical formations 
INFO[11-19|16:48:30] checking postgres 
INFO[11-19|16:48:30] found running postgres leader 
INFO[11-19|16:48:30] found 2 running postgres instances 
INFO[11-19|16:48:30] getting postgres status 
INFO[11-19|16:48:30] postgres claims to be read-write 
INFO[11-19|16:48:30] checking for running controller API 
INFO[11-19|16:48:30] found running controller API instances   n=2
INFO[11-19|16:48:30] found controller instance, checking critical formations 
INFO[11-19|16:48:30] scheduler looks up, moving on 
INFO[11-19|16:48:30] cluster fix complete 
```
Here is the debug gist:
https://gist.github.com/anonymous/f7d4d7cfcb8d2fcc9301

I noticed there appear to be some SQL errors in there to do with the blob store.
E..g
```
2015/11/19 06:25:01 discoverd: heartbeat blobstore (100.100.23.6:80) failed: unknown_error: Something went wrong
2015/11/19 06:25:11 error: ERROR: cannot execute INSERT in a read-only transaction (SQLSTATE 25006)
```

flynn version is: v20151104.1

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request moves the heartbeats out of the raft write path and persists them only in-memory on the discoverd leader. The leader is responsible for expiring service instances that do not heartbeat within their TTL.

Leaders cannot begin expiring instances until after they have had leadership for at least 2x TTL.

### Other notes

The vendored version of Bolt was really old and included a bug that caused a panic when shutting down Raft prematurely. It wasn't a safety issue but the panic was showing up in tests and was causing problems so I upgraded to the latest commit.

Closes #2158
Closes #2168
Closes #2073

### CODE REVIEW

**Comment by titanous** [comment]
Please split the bolt update into a separate commit.

File: discoverd/server/store.go [hunk 1]
```diff
@@ -475,20 +532,20 @@ func (s *Store) applyRemoveInstanceCommand(cmd []byte) error {
 	}
 
 	// Remove instance data.
-	entry := s.data.Instances[c.Service][c.ID]
+	inst := s.data.Instances[c.Service][c.ID]
 	delete(s.data.Instances[c.Service], c.ID)
```
Comment (titanous) [inline discoverd/server/store.go]: Delete from `s.heartbeats`?

**Comment by titanous** [comment]
Looks like the commit needs a DCO sign-off before CI will run.

**Comment by titanous** [comment]
This looks great!

**Comment by josephglanville** [comment]
When trying to bootstrap with these changes I am seeing a ton of instance expiries:
https://gist.github.com/c1044cfe2cece0b87b38
https://gist.github.com/841a00a7072cec6fb47e

File: discoverd/server/store.go (hunk 1 above)
Comment (benbjohnson) [inline discoverd/server/store.go]: Good catch, fixed in  3384896.

**Comment by benbjohnson** [comment]
@titanous I split the commit into bolt upgrade (a3d873a) and heartbeating (3384896). I must have killed the DCO line when I rebased last time. I added the DCO to both commits.

**Comment by benbjohnson** [comment]
@josephglanville `AddInstance()` wasn't returning an `ErrNotLeader` error for in-memory checks which was causing it to not redirect in the handler. Fixed here:

https://github.com/benbjohnson/flynn/commit/0c6f14ad9aacd28a67c8d459e551193349885d37#diff-e92f931b3419848b18a7209332fe268dR427

**Comment by titanous** [comment]
The bolt update commit message should have a `Godeps: ` subsystem prefix.

**Comment by benbjohnson** [comment]
@titanous I fixed the commit line. I'm just trying to figure out why the builds are failing now. It's showing 503s on different tests. Could it be caused by the longer leader election times from the 2x TTL?

**Comment by titanous** [comment]
@benbjohnson I think that was a fluke or a timeout that is too short, we can fix after merging. I've got another run going.

**Comment by titanous** [comment]
LGTM

**Comment by thomasfedb** [comment]
When can I expect to see this change in a stable release?

**Comment by titanous** [comment]
This weekend.

**Comment by thomasfedb** [comment]
Excellent! :sparkles:

**Pull request merged**